sparrow-wifi: init at 0-unstable-2024-07-25

dani0854 opened the pull request:
## Description of changes

https://github.com/ghostop14/sparrow-wifi

Fix #347468

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

dani0854 committed:
sparrow-wifi: init at 0-unstable-2024-07-25

dani0854 commented:
Result of `nixpkgs-review pr 259584` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>sparrow-wifi</li>
  </ul>
</details>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4098

FliegendeWurst commented on pkgs/by-name/sp/sparrow-wifi/package.nix:
```diff
@@ -0,0 +1,151 @@
+{ lib
+, python3
```
Please reformat with the new formatter.

FliegendeWurst commented on pkgs/by-name/sp/sparrow-wifi/package.nix:
```diff
@@ -0,0 +1,151 @@
+    })
+  ];
+
+  meta = with lib; {
+    homepage = "https://github.com/ghostop14/sparrow-wifi";
+    description = "Next-Gen GUI-based WiFi and Bluetooth Analyzer for Linux";
+    mainProgram = "sparrow-wifi";
+    license = licenses.gpl3;
+    maintainers = with maintainers; [ dsuetin ];
+    platforms = platforms.linux;
```
```suggestion
  meta = {
    homepage = "https://github.com/ghostop14/sparrow-wifi";
    description = "Next-Gen GUI-based WiFi and Bluetooth Analyzer for Linux";
    mainProgram = "sparrow-wifi";
    license = lib.licenses.gpl3;
    maintainers = with lib.maintainers; [ dsuetin ];
    platforms = lib.platforms.linux;
```

FliegendeWurst commented on pkgs/by-name/sp/sparrow-wifi/package.nix:
```diff
@@ -0,0 +1,151 @@
+
+  postInstall = ''
+    # Generate and install icon files
+    for size in 16 32 48 64 72 96 128 192 512 1024; do
```
I think it's sufficient to install one icon size.

FliegendeWurst commented on pkgs/by-name/sp/sparrow-wifi/package.nix:
```diff
@@ -0,0 +1,151 @@
+    elasticsearch
+    elastic-transport
+    pytz
+  ] ++ lib.singleton qt5.qtwayland;
```
```suggestion
  ] ++ [ qt5.qtwayland ];
```

dani0854 commented:
Also did some refactoring, and moved wrapping to `postFixup`, so I could use `program_PATH` and `program_PYTHONPATH`.

And added a patch from kali repositories to fix python 312 warnings about regexp.

dani0854 commented:
Since sparrow-wifi requires to be launched as root including GUI, in order for the desktop file to properly work I also added pkexec wrapper, that passes required variables for GUI to work on X11 and Wayland.

I tried creating policy files with `org.freedesktop.policykit.exec.allow_gui=true`, but this only passes X11 required variables through pkexec.

FliegendeWurst commented on pkgs/by-name/sp/sparrow-wifi/package.nix:
```diff
@@ -0,0 +1,179 @@
+in
+python3.pkgs.buildPythonApplication {
+  pname = "sparrow-wifi";
+  version = "unstable-2024-07-25";
```
```suggestion
  version = "0-unstable-2024-07-25";
```

dani0854 replied:
Done

FliegendeWurst commented on pkgs/by-name/sp/sparrow-wifi/package.nix:
```diff
@@ -0,0 +1,179 @@
+
+  postPatch = ''
+    substituteInPlace sparrow-wifi.py \
+      --replace-fail "/usr/bin/xgps" "${gpsd}/bin/xgps"
```
```suggestion
      --replace-fail "/usr/bin/xgps" "${lib.getExe' gpsd "xgps"}"
```
Same below.

dani0854 replied:
Done

FliegendeWurst commented on pkgs/by-name/sp/sparrow-wifi/package.nix:
```diff
@@ -0,0 +1,179 @@
+  sparrow-wifi-desktop = makeDesktopItem {
+    name = "sparrow-wifi";
+    desktopName = "Sparrow WiFi";
+    exec = "@out@/bin/sparrow-wifi-pkexec";
```
```suggestion
    exec = "sparrow-wifi-pkexec";
```
#308324

dani0854 replied:
Done

FliegendeWurst commented on pkgs/by-name/sp/sparrow-wifi/package.nix:
```diff
@@ -0,0 +1,179 @@
+    substituteAll ${sparrow-wifi-pkexec} $out/bin/sparrow-wifi-pkexec
+    chmod +x $out/bin/sparrow-wifi-pkexec
+
+    # add desktop file
+    mkdir -p $out/share/applications
+    substituteAll ${sparrow-wifi-desktop}/share/applications/sparrow-wifi.desktop $out/share/applications/sparrow-wifi.desktop
```
```suggestion
```
Use copyDesktopItems instead.

dani0854 replied:
Done

FliegendeWurst commented on pkgs/by-name/sp/sparrow-wifi/package.nix:
```diff
@@ -0,0 +1,179 @@
+python3.pkgs.buildPythonApplication {
+  pname = "sparrow-wifi";
+  version = "0-unstable-2024-07-25";
+  format = "other";
```
```suggestion
  pyproject = false;
```
Is preferred and has the same effect if I remember correctly.

dani0854 replied:
Done, I thought that it will default to setuptools, but `false` does indeed default to `other`.

FliegendeWurst commented on pkgs/by-name/sp/sparrow-wifi/package.nix:
```diff
@@ -0,0 +1,179 @@
+    john
+    iw
+    ubertooth
+    qt5.qtwayland
```
qtwayland would go in buildInputs, unless I am missing something.
Otherwise the result will depend on all development libraries.

dani0854 replied:
Done, thanks for pointing it out

FliegendeWurst commented on pkgs/by-name/sp/sparrow-wifi/package.nix:
```diff
@@ -0,0 +1,179 @@
+    aircrack-ng
+    john
+    iw
+    ubertooth
```
The path to ubertooth is hardcoded in the output. Probably you can omit the package here. Same for the others.

dani0854 replied:
Unfortunately the full path for executables that I am patching bellow is only used to check if executable is available:
https://github.com/ghostop14/sparrow-wifi/blob/b6018c5dfead5294cc2f810d34347a5b5f55aeb0/sparrowbluetooth.py#L1049C1-L1049C119

But when it actually launches it, it uses the one from PATH:
https://github.com/ghostop14/sparrow-wifi/blob/b6018c5dfead5294cc2f810d34347a5b5f55aeb0/sparrowbluetooth.py#L511

FliegendeWurst commented on pkgs/by-name/sp/sparrow-wifi/package.nix:
```diff
@@ -0,0 +1,179 @@
+  postFixup = ''
+    makeWrapperArgs+=(
+      --chdir "$out/lib/sparrow-wifi/"
+      --prefix PATH ':' "$program_PATH"
+      --prefix PYTHONPATH ':' "$program_PYTHONPATH"
```
I don't understand this wrapper, where are program_PATH and program_PYTHONPATH set?

dani0854 replied:
They are set by python wrap script during fixup phase, by going through `/nix-support/propagated-build-inputs` recursively.
https://github.com/NixOS/nixpkgs/blob/3a931f3d8027d7390439196237b2cd433b4b6e6c/pkgs/development/interpreters/python/wrap.sh#L105

Unfortunately the wrap phase only wraps executables installed in `bin`, and I can't copy them to `bin` since there are relative dependencies (which aren't executables), and no install or setup script is present, just some python files in a repository. And I didn't want to reimplement the recursive logic to generate PATH and PYTHONPATH for wrapper.

I know it's probably not the best to depend on those variables, since they are kind of internal logic of python wrap (or at least there isn't any documentation about them), also they are used outside of it multiple place in nixpkgs currently.

The other method would be to create custom `setup.py` and inject it, but I thought the approach of copying to `lib` and then manually wrapping it is simpler.

Do you have any other ideas of how it could be done?

FliegendeWurst replied:
I would try: `ln -s "$out"/lib/sparrow-wifi/sparrowwifiagent.py "$out"/bin/sparrowwifiagent`
After adding `(python3.withPackages (ps: dependencies))` to `buildInputs`. That way the scripts will have the right shebang, hopefully.
